cloud_storage: implement retention.ms/bytes for remote data

VladLazar opened the pull request:
## Cover letter
This PR introduces retention for cloud storage data. This implemented via a periodic house-keeping job that runs within
every `ntp_archiver` instance. The housekeeping job performs two task sequentially:
1. Advances the start offset of all manifests such that the remaining segments meet the retention policy specified for the partition.
`ntp_archiver` simply computes the offset and then issues a command to the archival STM to persist and replicate the new offset.
2. Run garbage collection. This removes all the segments that can be no longer queried by clients from cloud storage: segments below the current start offset in the manifest and segments that have been replaced with their compacted equivalent.

Once deletion from cloud storage is done, the metadata for the partition is updated and replicated.

## Backport Required

<!-- Specify which branches this should be backported to, e.g.: -->
- [X] not a bug fix
- [ ] issue does not exist in previous branches
- [ ] papercut/not impactful enough to backport
- [ ] v22.2.x
- [ ] v22.1.x
- [ ] v21.11.x

## UX changes
Partitions of topics enabled for remote storage now follow the topic's retention policy specified via retention.bytes and retention.ms. If those are unspecified, cluster level defaults are used. The migrations of existing topics are done in such a way that they preserve the previous behaviour (i.e. data will not be deleted from cloud storage for topics created before v22.3). However, note that for new topics retention is applied and data can be removed from cloud storage.

## Release notes
* Partitions of topics enabled for remote storage now follow the topic's retention policy specified via retention.bytes and retention.ms. If those are unspecified, cluster level defaults are used. The migrations of existing topics are done in such a way that they preserve the previous behaviour (i.e. data will not be deleted from cloud storage for topics created before v22.3). However, note that for new topics retention is applied and data can be removed from cloud storage.

VladLazar commented:
Changes in [force-push]:
Cherry picked https://github.com/redpanda-data/redpanda/commit/2c7a4e7caf94f5848b5480814be63196bda28ad9 to get the CI going.

jcsp commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +858,97 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+          _metadata_sync_timeout, _cloud_storage_initial_backoff, &_rtcnode);
+        retry_chain_logger ctxlog(archival_log, fib, _ntp.path());
+
+        // If deletion fails and the subsequent command to cleanup metadata
```
We probably want roughly the same error handling here as in remote_partition's deletion code: timeouts are transient + we should retry later, everything else should prompt us to abandoned the object in S3.

jcsp commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +858,97 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->truncate(
```
Do you have a good intuition for how this passes through into remote_partition, when the manifest is updated and stuff gets removed from remote_partition::_segments?

I wonder if it is possible to construct a unit test that intentionally instantiates a reader in flight while then calling into truncate()

VladLazar replied:
Good question! Not really. I had a look at the code though. It looks like we lazily update `remote_partition::_segments` from the manifest via `remote_partition::update_segments_incrementally` when making a reader. 

Currently, that only happens if the number of segments in the manifest is larger than the size of `remote_partition::_segments`. The Kafka layer is not aware that we have advanced the start offset (because `remote_partition::_segments` didn't sync with the manifest), so if we get a request below the start offset in the manifest we will try to download the corresponding segment. This will fail as the segment was removed from S3.

Why do we even have `remote_partition::_segments` and not use the manifest as the source of truth? That would make things a lot easier I think.

VladLazar replied:
Summarising a discussion from today:
* We can and should use the segments in the manifest as our source of truth. There's some complexity though as remote partitions adjust the offsets to Kafka offsets.
* We should add a unit test for the scenario described in my reply. It will fail for now, but it should start passing when changes to remote partition are made.

jcsp reviewed (commented):
This all looks mercifully neat and straightfoward.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/8423c539f32e457bfe11f53a4144109c70036925..2092387bb5de98294853f98d0c9afe5d7a4c31df) and [force-push](https://github.com/redpanda-data/redpanda/compare/2092387bb5de98294853f98d0c9afe5d7a4c31df..4fcd30f9286656b0e1da28f3feeabb05701f4d62):
* Rebased on dev to pick up the archival metadata STM changes
* Hook up the housekeeping loop into the archiver
* Added some retention metrics

jcsp commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -81,6 +85,14 @@ ntp_archiver::ntp_archiver(
     }
+
+    // Only run the housekeeping task for non-read replicas
+    if (_partition && !_partition->is_read_replica_mode_enabled()) {
```
Maybe this should be in run_upload_loop:
- Usually things like spawning background work isn't done in constructors (prefer ::start methods or in this case the run loop method)
- If it's spawned on the same path as the rest of the upload machinery, it can share the upgrade handling in https://github.com/redpanda-data/redpanda/pull/6913

VladLazar replied:
Yep, housekeeping moved into the upload loop now.

Lazin commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +872,136 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+        auto sync_timeout = config::shard_local_cfg()
+                              .cloud_storage_metadata_sync_timeout_ms.value();
+        auto deadline = ss::lowres_clock::now() + sync_timeout;
+        auto error = co_await _partition->archival_meta_stm()->cleanup_metadata(
```
I think that there is a race condition here. Between the initialization of the `to_remove` collection and the call to `cleanup_metadata` there are a lot of async calls and it can take awhile to delete everything in `to_remove`. If during deletion the `start_offset` will be moved forward or some segments will be replaced, the metadata of those segments will be lost. I think that the garbage collection should block uploads. Or alternatively, it should be called from the upload loop. The houskeeping loop could arrange that somehow (with the flag or a semaphore).

Lazin replied:
I can see the merit in having a separate fiber that runs retention especially when we want upload and retention to have different priorities.

VladLazar replied:
The `start_offset` is only advanced by house keeping so that should not be an issue. That leaves the reuploading of compacted segments which is indeed an issue. Could we just block the re-upload of compacted segments while GC is running? This leaves normal uploads untouched.

jcsp replied:
I would be +1 on making retention updates, uploads and re-uploads all mutually-exclusive in time.  Makes it so much easier to reason about and test.

VladLazar replied:
Done. I moved the housekeeping into the upload loop. I feel like this is a bit of a big hammer. We could have just blocked compacted uploads, but let's come back to it after the release.

Lazin commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +872,136 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+
+        // If deletion fails and the subsequent command to cleanup metadata
+        // succeeds, the segment will become orphaned.
+        auto res = co_await _remote.delete_object(
```
There should be some audit trails for segment deletion on info level. If we will ever have a data loss in the cloud storage the retention and GC will be a prime suspect. We will need a way to tell when something is deleted (or most importantly not deleted) by the retention without bumping logs to debug or trace.

jcsp replied:
+1, these are very verbose messages but worth having.  It will be the same situation as local storage logs, where by default we log things like starting a new segment, but for high scale systems we provide advice (https://vectorizedio.atlassian.net/wiki/spaces/CORE/pages/99844097/Handling+Large+Partition+Counts+in+Redpanda+22.2#Logging) on which subsystems to tune down to debug level to manage log volume.

I will update that scale wiki page to mention the cloud storage loggers too.

VladLazar replied:
Makes sense to me. I'll add some logging that includes the segment names at info level. Perhaps in the future we can have some in memory store that can be queried via `rpk audit` (or something similar).

VladLazar replied:
Done. All segment deletions are logged at the info level with the remote path.

Lazin reviewed (commented):
Just a few comments. Great progress overall!

Lazin commented on src/v/cloud_storage/partition_manifest.cc:
```diff
@@ -300,6 +300,20 @@ partition_manifest::const_reverse_iterator partition_manifest::rend() const {
 size_t partition_manifest::size() const { return _segments.size(); }
 
+size_t
+partition_manifest::cloud_log_size(std::optional<model::offset> end) const {
```
if the goal is to calculate total size (including the segments that has to be removed) then you need to take `_replaced` into account.

VladLazar replied:
Hmm. Good point about the start-offset. Functionally, I don't think it makes a difference. This function gets called only when applying retention, and at that point `_segments` should have been trimmed. I'll make the change for consistency though.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/4fcd30f9286656b0e1da28f3feeabb05701f4d62..6a8d71e4e59a0f005a9c64727f2a4db68eda567b):
* Tie the lifetime of the housekeeping to the upload loop
* Fix a couple error condition checks in archival STM
* Allow the cleanup_metadata command to proceed when there's no replaced segments, but the start offset was advanced and segments need to be removed from the manifest.

VladLazar commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +872,136 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this] {
+                          // TODO(vlad): Perhaps use a different
+                          // scheulding group here.
+                          return ss::with_scheduling_group(_upload_sg, [this] {
```
I moved the housekeeping into the upload loop, so now it makes sense to use the same scheduling group,

VladLazar commented:
Changes in last two force pushes (https://github.com/redpanda-data/redpanda/compare/7f03111e3f6639838b5a54e11f7909e9fdbaa894..c7fccc5d18be3007ba06ce1b62d0ba7f9d5c03c6 and https://github.com/redpanda-data/redpanda/compare/6a8d71e4e59a0f005a9c64727f2a4db68eda567b..7f03111e3f6639838b5a54e11f7909e9fdbaa894):
* Moved the housekeeping into the upload loop to have it be mutually exclusive with uploads
* Audit logging for removal of segments
* Ducktape tests for retention via metrics

VladLazar commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +858,97 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+
+        // If deletion fails and the subsequent command to cleanup metadata
+        // succeeds, the segment will become orphaned.
+        co_await _remote.delete_object(_bucket, s3::object_key{path}, fib);
```
Missed the comment. Will do on the next push. I'd like to get a CI round in.

VladLazar replied:
Done.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/c7fccc5d18be3007ba06ce1b62d0ba7f9d5c03c6..1a85584a4e4a2214957e7dc0eb529e7191b82a79):
* make deletion loop concurrent
* start at `start_offset` when computing cloud log size

Lazin commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +871,135 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+
+    std::atomic<size_t> successful_deletes{0};
+    ssx::semaphore limiter{_concurrency, "archival/delete"};
+    for (const auto& path : to_remove) {
```
nit: check abort source in the loop, otherwise redpanda may not stop in timely manner

VladLazar replied:
Done.

Lazin commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +871,135 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+    std::atomic<size_t> successful_deletes{0};
+    ssx::semaphore limiter{_concurrency, "archival/delete"};
+    for (const auto& path : to_remove) {
+        auto units = co_await ss::get_units(limiter, 1);
```
I don't think that this achieves the desired goal. The loop will still be sequential. I think that you might want to use `ss::coroutine::all` or `ss::max_concurrent_for_each` here.

VladLazar replied:
Yep. You're right. Doesn't actually start futures in parallel. Thanks for pointing it out.

VladLazar commented on src/v/cloud_storage/partition_manifest.cc:
```diff
@@ -300,6 +300,13 @@ partition_manifest::const_reverse_iterator partition_manifest::rend() const {
 size_t partition_manifest::size() const { return _segments.size(); }
 
+size_t partition_manifest::cloud_log_size() const {
+    return std::transform_reduce(
```
It does take `_start_offset` into account now by starting from the segment which begins on `start_offset`.

> If "all segments" is actually "all segments available to the user" then it should count everything above the _start_offset.

Yes, that's the intention. I'll update the comment in the header file a bit.

VladLazar replied:
Updated the comment.

VladLazar commented on src/v/archival/retention_calculator.h:
```diff
@@ -0,0 +1,95 @@
+ * that timestamps are monotonically increasing. That's not always
+ * the case, but treating them as such works well enough in most cases.
+ */
+class time_based_strategy final : public retention_strategy {
```
Done.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/1a85584a4e4a2214957e7dc0eb529e7191b82a79..9530385ef472e79ee410d86c451335693a8ae665):
* Fix deletion concurrency
* Redo the metrics. Previously, a lot of partition level metrics were exposed on `public_metrics`. That increases cardinality by too much, so I just added a couple topic level metrics.

VladLazar commented:
Pushed again as CI was stuck.

VladLazar commented:
Failure is `PandaProxyBasicAuthTest.test_fetch_topic`. The test doesn't have cloud storage enabled, so I very much doubt that it's a regression introduced by these changes. I asked Ben about it and will create an issue if it's something new.

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/55afda0eefcec1c2a331b7ef3824911a2b447b3e..744b6bb8fe311c4a5257003cf0978cdeb7ba86f7):
* Use `archival_metadata_stm::get_segments_to_cleanup` instead of duplicating that logic.

jcsp commented on src/v/archival/probe.cc:
```diff
@@ -67,6 +74,40 @@ ntp_level_probe::ntp_level_probe(
+      partition_label(ntp.tp.partition()),
+    };
+
+    auto aggregate_labels = std::vector<sm::label>{
```
We might want to revisit this at some point to maintain some shard-wide counters, rather than doing per-ntp counters that will get summed up at scrape time.

jcsp commented on src/v/archival/probe.h:
```diff
@@ -52,8 +65,14 @@ class ntp_level_probe {
+    /// Number of segments deleted by garbage collection
+    int64_t _segments_deleted = 0;
+    /// Current start offset of the manifest
+    model::offset _start_offset{0};
```
This seems to be unused (+ we probably don't want it in public_metrics due to cardinality and the fact that it's a raft offset that doesn't make sense to the user)

VladLazar replied:
Correct. I missed removing it when I re-did the metrics bit.

VladLazar replied:
Done.

jcsp reviewed (commented):
Minor comments only + this needs a rebase because it raced with Michal's pr on the test file.

Otherwise good to merge tomorrow.

VladLazar commented on src/v/cloud_storage/partition_manifest.h:
```diff
@@ -172,6 +172,11 @@ class partition_manifest final : public base_manifest {
+    // Computes the size in bytes of all segments available to clients
+    // (i.e. all segments after and including the segment that starts at
+    // the current _start_offset).
+    size_t cloud_log_size() const;
```
Good point. Will update.

VladLazar replied:
Fixed.

Lazin reviewed (dismissed):
LGTM

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/744b6bb8fe311c4a5257003cf0978cdeb7ba86f7..989d5eb0b187998f2e2d00f640cbbe2afaef0306):
* Use uint64_t for cloud log size
* Remove unused _start_offset from probe

VladLazar commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/989d5eb0b187998f2e2d00f640cbbe2afaef0306..66c835795fca4f8a0b882c75c05d482d9288e5b9):
* rebased on dev pre-merge

dotnwat commented on src/v/archival/ntp_archiver_service.cc:
```diff
@@ -854,4 +871,137 @@ ntp_archiver::maybe_truncate_manifest(retry_chain_node& rtc) {
+
+    auto to_remove = _partition->archival_meta_stm()->get_segments_to_cleanup();
+
+    std::atomic<size_t> successful_deletes{0};
```
@VladLazar i don't see any reason you would need an atomic here. is there something i'm not seeing? if not, can you change this to a plain `size_t` since std::atomic will blow cache lines...

VladLazar replied:
`succesful_deletes` can be incremented concurrently by the futures spawned by `ss::max_concurrent_for_each` below. I could probably rework this code to return a list of futures and infer how many deletions where successful from that.

dotnwat replied:
> succesful_deletes can be incremented concurrently by the futures spawned by ss::max_concurrent_for_each below. I could probably rework this code to return a list of futures and infer how many deletions where successful from that.

how so? even if multiple concurrent continuations are runnable, only one at a time will run on the same core. are you saying there is physical concurrency across cores here?

VladLazar replied:
You're right. Thanks for pointing it out. There's no physical concurrency here. The word "parallel" in the docstring made me think that `max_concurrent_for_each` goes across cores. Looking at the code, it doesn't.
